### PR TITLE
ARM64: Add support for AM6x A53 pinctrl

### PR DESCRIPTION
**Pull request by gramsay0**
Add support for:

- TI K3 pinctrl
- AM6x A53 pinctrl
- NS16550 serial driver pinctrl

Demonstrate on UART0 of the phyCORE-AM62x A53 board.

"K3" is a common architecture used between different TI processor families

### CODE REVIEW

File: soc/arm64/ti_sitara/am6x/mmu_regions.c
Comment (gramsay0) [inline soc/arm64/ti_sitara/am6x/mmu_regions.c:12]: This is a bit gross...

The pinctrl registers are only 0x2ac (684) bytes.
The MMU minimum granularity is 4kB.

The arm64 MMU driver seems to ignore this configuration unless it is rounded up to a page size.
I'm not sure if this is the correct fix, or if this "rounding up" should be done in the MMU macros or driver code..

File: drivers/pinctrl/pinctrl_ti_k3.c
Comment (SgrrZhf) [inline drivers/pinctrl/pinctrl_ti_k3.c:7]: It looks like this macro isn't used here

Comment (gramsay0) [reply drivers/pinctrl/pinctrl_ti_k3.c]: Fixed

File: drivers/pinctrl/pinctrl_ti_k3.c
Comment (carlocaione) [inline drivers/pinctrl/pinctrl_ti_k3.c:17]: This is really something to avoid for MMU-enabled platforms. You are reading and writing a physical address retrieved from the DT implicitly assuming that you have a 1:1 mapping between phys and virt addresses and it is not always the case (as long as you force it).

The MMIO region of the pinctlr driver should map the MMIO space using the device MMIO APIs (see `include/zephyr/sys/device_mmio.h`) and the driver should access the virtual address instead.

File: soc/arm64/ti_sitara/am6x/mmu_regions.c
Comment (carlocaione) [reply soc/arm64/ti_sitara/am6x/mmu_regions.c:12]: it's really odd having a MMIO space that is not 4k aligned. What else is in that page besides pinctlr?

File: soc/arm64/ti_sitara/am6x/mmu_regions.c
Comment (carlocaione) [inline soc/arm64/ti_sitara/am6x/mmu_regions.c:29]: No, we are trying to move away from forcing a 1:1 mapping. The pinctrl should use the device MMIO APIs to dynamically map the memory region, see for example how this is done for `drivers/serial/uart_pl011.c`

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: soc/arm64/ti_sitara/am6x/mmu_regions.c
Comment (gramsay0) [reply soc/arm64/ti_sitara/am6x/mmu_regions.c:12]: The PADCFG register block is actually 32kB:
image
Although only a small section of this block is used for pinctrl.

Here's a dtsi from Linux:
https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/ti/k3-am62-main.dtsi#L187

Comment (gramsay0) [reply soc/arm64/ti_sitara/am6x/mmu_regions.c]: After converting to use MMIO this is no longer an issue.
`device_map` calls `k_mem_region_align` which rounds up to an MMU page size

File: drivers/pinctrl/pinctrl_ti_k3.c
Comment (gramsay0) [reply drivers/pinctrl/pinctrl_ti_k3.c]: Fixed, thanks for the feedback!

File: drivers/pinctrl/pinctrl_ti_k3.c
Comment (carlocaione) [inline drivers/pinctrl/pinctrl_ti_k3.c:8]: move this after the includes.

**Review by carlocaione — DISMISSED** [review]
Acking the architectural part. For the pinctrl part I'll leave this to @gmarull

File: drivers/pinctrl/pinctrl_ti_k3.c
Comment (gramsay0) [reply drivers/pinctrl/pinctrl_ti_k3.c]: Fixed

**Review by carlocaione — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Comment by gmarull** [comment]
Note: PR was opened on 28/02, merged 2/03, so after just **2 days**. Can we please establish some rules like require maintainer approval, and only merge if the maintainer is not responsive for e.g. 3–4 weeks?

**Comment by nashif** [comment]
> Note: PR was opened on 28/02, merged 2/03, so after just **2 days**. Can we please establish some rules like require maintainer approval, and only merge if the maintainer is not responsive for e.g. 3–4 weeks?

Agree we need to let something like this sit for longer before merging. The PR was not assigned to anyone, maybe it should have assigned manually to get proper review before merging, however, who is the maintainer of a driver instance or a board is not going to be clear. Maintainer of the a subsystem does not need to approve every instance for something to go in, this falls on the SoC or board maintainer.

Reverting is a probably extreme in this case, give a chance for someone to fix issues before reverting or submit a fix if it is straightforward. So unless it is REALY broken, I would not go an revert everything i did not get a chance to review or if I diagree with certain apects of a merge PR.

**Comment by gmarull** [comment]
> > Note: PR was opened on 28/02, merged 2/03, so after just **2 days**. Can we please establish some rules like require maintainer approval, and only merge if the maintainer is not responsive for e.g. 3–4 weeks?
> 
> Agree we need to let something like this sit for longer before merging. The PR was not assigned to anyone, maybe it should have assigned manually to get proper review before merging, however, who is the maintainer of a driver instance or a board is not going to be clear. Maintainer of the a subsystem does not need to approve every instance for something to go in, this falls on the SoC or board maintainer.
> 
> Reverting is a probably extreme in this case, give a chance for someone to fix issues before reverting or submit a fix if it is straightforward. So unless it is REALY broken, I would not go an revert everything i did not get a chance to review or if I diagree with certain apects of a merge PR.

Well, it is a really broken implementation. It compiles, sure, but it has implemented pin control in a custom way (see my comment about props above), something we spent a long time standardizing and documenting as a requirement. Changes in the uart16550 driver are also wrong. We should be able to revert as fast as we merge in cases like this; otherwise, we're creating the incentive to get approvals and merge fast because reverts are difficult/impossible.

**Comment by carlocaione** [comment]
> Maintainer of the a subsystem does not need to approve every instance for something to go in, this falls on the SoC or board maintainer.

Shouldn't this be the opposite? I mean, as I tried to point out here https://github.com/zephyrproject-rtos/zephyr/pull/55245#pullrequestreview-1319216013 I can very well be familiar with the SoC or board but being clueless about the subsystem.

**Comment by nashif** [comment]
> Shouldn't this be the opposite? I mean, as I tried to point out here https://github.com/zephyrproject-rtos/zephyr/pull/55245#pullrequestreview-1319216013 I can very well be familiar with the SoC or board but being clueless about the subsystem.

In this case you will have to solicit reviews and feedback from the experts or the maintainers of areas being touched.

**Pull request merged**
